_summaries/katharopoulos2020transformers_1.md

```diff
@@ -0,0 +1,56 @@
+---
+layout: summary
+title: Summary
+paper: katharopoulos2020transformers
+# Please fill out info below
+author: kelseyball
+score: 9
+---
+
+# Core Idea
+Expressing self-attention using kernel feature maps can reduce its time and space complexity from quadratic to linear
+
+# Main Contributions
+- Formulation of "linear attention" using kernel feature maps
+- Formulation of autoregressive transformers with linear time, constant memory (can also be expressed as RNNs)
+- Experiments demonstrating speedups of autoregressive linear transformers over reformer and transformer with softmax attention
+
+# Softmax Attention to Linear Attention via Kernel Feature Maps
+
+The transformer from Vaswani et. al. 2017 implements the following form of self-attention, where the similarity score is the exponential of the dot product between a query and a key:
+
+<img width="500px" src="katharopoulos2020transformers_1a.png"/>
+
+Generalizing from dot product to any positive-valued similarity function, we can rewrite (2) as:
+
+<img width="500px" src="katharopoulos2020transformers_2a.png"/>
+
+Our similarity function can also be a **kernel function**. In this context, a kernel function takes two vectors, maps them to another vector space using a **kernel feature map**, and returns their inner product in that space. Here is the self-attention equation using kernel feature map *phi*:
+
+<img width="500px" src="katharopoulos2020transformers_3a.png"/>
+
+The associative property of matrix multiplication allows us to make the following simplification,
+
+<img width="500px" src="katharopoulos2020transformers_4a.png"/>
+
+The reduction from quadratic to linear complexity comes from the fact that the matrix product of the keys and values can be computed once and reused for every query.
+
+# Linear Autoregressive Transformers
+- In autoregressive tasks, inputs from future timesteps are masked to the model. For autoregressive transformers, we cannot precompute the full value of the key-value matrix product as shown above.
+- However, the authors show that part of the attention computation at each timestep can be computed iteratively. In other words, it can be computed in constant time using the previous value. This observation allows for time complexity linear in the sequence length.
```

Comment:
I think you should bring up these two points for some context. With autoregressive transformers, it is easier to parallelize during training and not possible during test time. The attention computation must also be repeated for previous time steps.